source/darwin/nw_socket.c

```diff
@@ -0,0 +1,2052 @@
+            if (hostname != NULL) {
+                size_t hostname_len = strlen(hostname);
+                size_t buffer_size = AWS_ARRAY_SIZE(socket->local_endpoint.address);
+                size_t to_copy = aws_min_size(hostname_len, buffer_size);
+                memcpy(socket->local_endpoint.address, hostname, to_copy);
+                socket->local_endpoint.port = port;
+            }
+            nw_release(local_endpoint);
+
+            AWS_LOGF_TRACE(
+                AWS_LS_IO_SOCKET,
+                "id=%p handle=%p: set local endpoint %s:%d",
+                (void *)socket,
+                socket->io_handle.data.handle,
+                socket->local_endpoint.address,
+                port);
+        }
+        s_set_socket_state(nw_socket, socket, CONNECTED_WRITE | CONNECTED_READ);
+        aws_mutex_unlock(&nw_socket->synced_data.lock);
+
+        nw_socket->connection_setup = true;
+        // Cancel the connection timeout task
+        if (nw_socket->timeout_args) {
+            aws_event_loop_cancel_task(nw_socket->event_loop, &nw_socket->timeout_args->task);
+        }
+        aws_ref_count_acquire(&nw_socket->external_ref_count);
+        s_schedule_on_connection_result(nw_socket, AWS_OP_SUCCESS);
+        s_schedule_next_read(nw_socket);
+        aws_ref_count_release(&nw_socket->external_ref_count);
+    } else if (state == nw_connection_state_waiting) {
+        AWS_LOGF_DEBUG(
+            AWS_LS_IO_SOCKET,
+            "id=%p handle=%p: socket connection is waiting for a usable network before re-attempting. ErrorCode: %d.",
+            (void *)nw_socket,
+            (void *)nw_socket->os_handle.nw_connection,
+            connection_args->error);
+    } else if (state == nw_connection_state_preparing) {
+        AWS_LOGF_DEBUG(
+            AWS_LS_IO_SOCKET,
+            "id=%p handle=%p: socket connection is in the process of establishing. ErrorCode: %d",
+            (void *)nw_socket,
+            (void *)nw_socket->os_handle.nw_connection,
+            connection_args->error);
+    } else if (state == nw_connection_state_failed) {
+        /* any error, including if closed remotely in error */
+        AWS_LOGF_ERROR(
+            AWS_LS_IO_SOCKET,
+            "id=%p handle=%p: socket connection failed with error code: %d.",
+            (void *)nw_socket,
+            (void *)nw_socket->os_handle.nw_connection,
+            connection_args->error);
+    }
+
+    if (connection_args->error) {
+        /* any error, including if closed remotely in error */
+        AWS_LOGF_ERROR(
+            AWS_LS_IO_SOCKET,
+            "id=%p handle=%p: socket connection get apple os error code: %d",
+            (void *)nw_socket,
+            (void *)nw_socket->os_handle.nw_connection,
+            connection_args->error);
+
+        int error_code = s_determine_socket_error(connection_args->error);
+        nw_socket->last_error = error_code;
+        aws_mutex_lock(&nw_socket->synced_data.lock);
```

Comment:
Trivial: There are a lot of places we use `aws_mutex_lock(&nw_socket->synced_data.lock)` despite having the helper functions `s_lock_socket_synced_data` and `s_unlock_socket_synced_data`.  We should either always use the helper functions or get rid of them entirely. A quick search shows 16 manual  calls on synced_data.lock and 8 of the helper function.